src/linear_algebra/eigenspace.lean

```diff
@@ -0,0 +1,242 @@
+/-
+Copyright (c) 2020 Alexander Bentkamp. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alexander Bentkamp.
+-/
+
+import field_theory.algebraic_closure
+import linear_algebra.finsupp
+
+/-!
+# Eigenvectors and eigenvalues
+
+This file defines eigenspaces and eigenvalues.
+
+An eigenspace of a linear map `f` for a scalar `μ` is the kernel of the map `(f - μ • id)`.
+The nonzero elements of an eigenspace are eigenvectors `x`. They have the property `f x = μ • x`.
+If there are eigenvectors for a scalar `μ`, the scalar `μ` is called an eigenvalue.
+
+## Notations
+
+The expression `algebra_map α (β →ₗ[α] β)` appears very often, which is why we
+use `am` as a local notation for it.
+
+## References
+
+* [Sheldon Axler, *Down with determinants!*,
+  https://www.maa.org/sites/default/files/pdf/awards/Axler-Ford-1996.pdf][axler1996]
+* https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors
+
+## Tags
+
+eigenspace, eigenvector, eigenvalue, eigen
+-/
+
+universes u v w
+
+open vector_space principal_ideal_ring polynomial finite_dimensional
+
+variables {α : Type v} {β : Type w} [add_comm_group β]
+
+local notation `am` := algebra_map α (β →ₗ[α] β)
+
+/-- The subspace `eigenspace f μ` for a linear map `f` and a scalar `μ` consists of all vectors `x`
+    such that `f x = μ • x`. -/
+def eigenspace [field α] [vector_space α β] (f : β →ₗ[α] β) (μ : α) : submodule α β :=
```

Comment:
The definitions of `eigenspace`, `eigenvector` and `eigenvalue` and a lot of the lemmas make sense for any module over a `comm_ring`, could you generalize them?

Reply:
I did. However, only one lemma generalized.